text/trace/0220-messaging-semantic-conventions-span-structure.md

```diff
@@ -0,0 +1,626 @@
+settled, when possible.
+
+## Proposed changes and additions to the messaging semantic conventions
+
+### Operation name
+
+The following operations related to messages are covered by these semantic
+conventions:
+
+| Operation name | Description |
+|----------------|-------------|
+| `publish`      | One or more messages are provided for publishing to an intermediary. |
+| `create`       | A message is created. |
+| `receive`      | One or more messages are requested by a consumer. |
+| `deliver`      | One or more messages are passed to a consumer. |
+| `settle`       | One or more messages are settled. |
+
+For further details about each of those operations refer to the [section about trace structure](#trace-structure).
+
+### Span kind
+
+[Span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind)
+SHOULD be set according to the following table, based on the operation a span describes.
+
+| Operation name | Span kind|
+|----------------|-------------|
+| `publish`      | `PRODUCER`, if no `create` spans are present. |
+| `create`       | `PRODUCER` |
+| `receive`      | `CONSUMER` |
+| `deliver`      | `CONSUMER` |
+
+The kind of `settle` spans should be set according to the [generic specification about span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind),
+e. g. it should be set to `CLIENT` if the `settle` spans models a synchronous call
+to the intermediary.
+
+Setting span kinds according to this table ensures that span links between
+consumers and producers always go from a `PRODUCER` span on the producer side to
+a `CONSUMER` span on the consumer side. This allows analysis tools to interpret
+linked traces without the need for additional semantic hints.
+
+### Span relationships
+
+#### Producer spans
+
+"Publish" spans SHOULD be created for operations of providing messages for
+sending or publishing to an intermediary. A single "Publish" span can account
+for a single message, or for multiple messages (in the case of providing
+messages in batches). "Create" spans MAY be created. A single "Create" span
+SHOULD account only for a single message. "Create" spans SHOULD either be
+children or links of the related "Publish" span.
+
+If a "Create" span exists for a message, its context SHOULD be injected into
+the message. If no "Create" span exists, the context of the related "Publish"
+span SHOULD be injected into the message.
+
+#### Consumer spans
+
+##### Push-based scenarios
+
+"Deliver" spans SHOULD be created for operations of passing messages to the
+application when those operations are not initiated by the application
+code.
+
+##### Pull-based scenarios
+
+"Receive" spans SHOULD be created for operations of passing messages to the
+application when those operations are initiated by the application code.
+
+##### General considerations
```

Comment:
It seems the spec skips on "process" spans. I think this is unfortunate, because while it is true that they cannot be "required" and not always produced, in some cases they can, and in probably quite a few cases the application owner might find it worth their while to even add a bit of manual code (maybe just an annotation with appropriate OTel support) as the per-single message processing is vital if you want to know e.g. which messages are taking longer than others to process.

For these cases it would be great if the spec here could also suggest how to set the span relationships.

A process span will often have two potential parents: The "deliver" or "receive" span (or maybe it's parent if such a span is not created) and the (single) Create/Produce span for the message that is processed. So there are three potential ways to capture this (using produce and deliver instead of produce/create and deliver/receive for simplicity):

1. All-links: Link both Produce and Deliver
2. Link Produce and use Deliver as parent (continue trace of Deliver): This emphasizes the technical execution flow of the code, and break the processing of a message (chain) into multiple traces while bundling the processing of the same batch of (potentially unrelated) messages into the same trace.
3. Use Produce as parent and link Deliver (continue trace of Produce): This emphasizes the information flow through the distributed system and will lead to the processing of a (chain of) messages to become a single trace. The technical execution flow will be in linked traces that only go up to the deliver/receive.

I like 3 more from a gut feeling, but there might be cases where 2 or 1 need to be used, e.g. when you have some kind of circular token passing, where a message flow never ends (did include such scenarios in your design considerations?).

Reply:
> EDIT: It would be OK to create a follow-up issue and mention it in the OTEP text, if you think this process span is particularly tricky.

It's tricky in the sense that we weren't able to come up with conventions that apply to "Process" spans in a general way.

We can generally say that we require a "Receive" spans for all pull-based consumer scenarios, and we can generally say that we require a "Deliver" spans for all push-based consumer scenarios. However, we can't say that we _generally_ require "Process" spans, as those might look different across scenarios. Sometimes they aren't created at all, sometimes there might be one for a batch of messages, sometimes there might be multiple "Process" spans for a single message, sometimes there might be multiple "Process" spans for the same batch ...

The goal of this OTEP is to provide a minimal foundation for instrumenting messaging scenarios, which provides consistency, but gives instrumentors freedom to flexibly add "Process" spans on top of it in ways that makes sense for their scenarios (there's the same aim for transport layer instrumentation).

I think there's limited use in aiming at describing and specifying all the ways in which "Process" spans can be created. Either we artificially limit instrumentor, or we come up with such a long list  that it defies the use of "normative" conventions.

All the three options you describe above are supported with the current draft of conventions in this OTEP (given that "Deliver" links to the "Create" context in all three cases, it can parent in addition). I think that's good and desirable and gives instrumentors flexibility. I think we shouldn't limit them to just one of those options.

Reply:
Discussed in the workgroup: we will add a follow-up issue and add it as an open issue in the OTEP.

Reply:
Over-restricting is probably not good here. But for example, the particular question I had in my comment, it might be helpful to add a recommendation at least, i.e. in the form "IF you face the choice between using X and Y as parent, it is conventional but not required to prefer Y.

Reply:
I submitted a separate issue regarding conventions for "Process" spans: https://github.com/open-telemetry/opentelemetry-specification/issues/3395

I added a reference to it under the "Future possibilities" section of this OTEP, as it's not a blocking issue for the content of this OTEP. However, we intend to start working on it in the very near term.